Clearancing improvements

cameronjacoby opened the pull request:
Includes the following improvements to the clearancing tool:
- UI for admins (warehouse workers) to scan multiple barcodes without touching the keyboard and clearance item id's in batch
- Ability for admins (merchandisers) to set the minimum clearance price on item types
- Ability for admins to pull different reports for themselves and vendors, as well as error reports on items not clearanced
- Clearance batch show page listing all clearanced items in the batch
- UI for filtering items by status and searching items by id, style name, or item type

nickflux commented on config/initializers/mime_types.rb:
```diff
@@ -2,3 +2,5 @@
 # Add new mime types for use in respond_to blocks:
 # Mime::Type.register "text/richtext", :rtf
+
+Mime::Type.register "application/xls", :xls
```
Why XLS rather than CSV? The returned files are CSV files. Might this set up an expectation of actual Excel files?

cameronjacoby replied:
I thought I needed this since I'm using "format.xls" to export reports. The reason I chose XLS is to make it easier for admins to format and/or manipulate data in the reports once exported -- I thought I'd save them the extra step of converting the files from CSV to XLS.

nickflux replied:
Fair enough.

nickflux commented on app/models/type.rb:
```diff
@@ -0,0 +1,9 @@
+class Type < ActiveRecord::Base
```
Probably best not to call this `type` since that is reserved by Rails and also very generic. At Stitch Fix we call it `item_type` which is better, but not much (since it refers to styles rather than items). `style_type` would be best, I think. Maybe change the column on the existing table and then you also can get rid of the `self.inheritance_column = :_type_disabled` line in `models/style`.

nickflux commented on app/controllers/clearance_batches_controller.rb:
```diff
@@ -1,25 +1,54 @@
+    end
+  end
+
+  def show_errors
```
This is not RESTful but it is a show action. My preference is to deviate from RESTful actions only with a *very* good reason. Maybe the route could be `/clearance_batch/1/errors` and have it all in the show action.

nickflux commented on app/controllers/clearance_batches_controller.rb:
```diff
@@ -1,25 +1,54 @@
 class ClearanceBatchesController < ApplicationController
+  before_action :set_clearance_batch, only: [:show, :show_errors]
```
:+1:

nickflux commented on app/controllers/items_controller.rb:
```diff
@@ -0,0 +1,28 @@
+    end
+  end
+
+  def search
```
Think about a `SearchService` class that you can just hand all the params too and return the items. If you wanted to extend this it would get out of hand quickly. Also, the `case` structure is not great - try using an instance of the service class to build up the search w/o it.

nickflux commented on app/models/clearance_batch.rb:
```diff
@@ -1,5 +1,13 @@
 
   has_many :items
+  serialize :clearancing_errors
+  after_initialize :set_clearancing_errors
```
What's this for? I would generally avoid callbacks outside of controllers. They do unexpected things.

cameronjacoby replied:
This was for setting the clearancing_errors column to an empty array after the clearance_batch was initialized. No longer needed, since I refactored clearancing_errors to be a separate table with an association rather than a column with data type array.

nickflux commented on app/models/clearance_batch.rb:
```diff
@@ -1,5 +1,13 @@
 class ClearanceBatch < ActiveRecord::Base
 
   has_many :items
+  serialize :clearancing_errors
```
What's the use case for storing the errors?

cameronjacoby replied:
My idea was that an admin could be clearancing thousands of items at once and may run into a case where there are enough errors (say, 100 or so) where she may want to save that information to look into later.

nickflux commented on app/models/item.rb:
```diff
@@ -1,17 +1,31 @@
   belongs_to :clearance_batch
 
-  CLEARANCE_PRICE_PERCENTAGE  = BigDecimal.new('0.75')
+  default_scope { order("id desc") }
```
NOOOOOOO! `default_scope` is the Devil! Like a lot of Rails magic it is very unpredictable.

nickflux commented on app/models/item.rb:
```diff
@@ -1,17 +1,31 @@
+    end
+  end
+
+  def margin
```
Did you used to work at Stitch Fix by any chance?

nickflux commented on app/services/clearance_batch_reports_service.rb:
```diff
@@ -0,0 +1,59 @@
+      "clearance_batch_id"
+    ]
+    if report_type == "vendor"
+      headers -= ["status", "wholesale_price", "retail_price", "margin", "style_id", "clearance_batch_id"]
```
Odd that this is subtractive. Probably better to set the minimum possible in the initial array and then add what else you need for a particular case. It's best not to surprise subsequent developers.

nickflux commented on app/services/clearance_batch_reports_service.rb:
```diff
@@ -0,0 +1,59 @@
+    CSV.generate(options) do |csv|
+      csv << headers
+      clearance_batch.items.each do |item|
+        values = [
```
Could you use the headers to build this up dynamically?

nickflux commented on app/services/clearancing_service.rb:
```diff
@@ -1,54 +1,51 @@
-      return "Item id #{potential_item_id} could not be clearanced"
+private
+
+  def add_items_to_clearance_batch(potential_item_ids, new_clearance_batch)
```
This method is too big. It feels a bit de-factored. Can you break it up into more understandable methods?

nickflux commented on config/routes.rb:
```diff
@@ -1,6 +1,17 @@
+
+  # clearance batches
+  resources :clearance_batches, only: [:index, :create, :show]
+  get "/clearance_batches/:id/errors", to: "clearance_batches#show_errors", as: "clearance_batch_errors"
```
OK, so you did this. But could you use the show action in the controller?

cameronjacoby replied:
I decided to go with a nested resource instead and use the index action in the clearancing_errors controller -- what do you think of that approach?

nickflux commented on config/routes.rb:
```diff
@@ -1,6 +1,17 @@
+
+  # types
+  resources :types, only: [:index, :edit, :update]
+  get "/settings", to: "types#index", as: "settings"
```
Why the naming mis-match?

cameronjacoby replied:
At the time I was thinking there would eventually be a more built-out settings page, so I was setting up for that. But for now, I think you're right -- just using the style_types_path makes more sense.

nickflux commented on config/routes.rb:
```diff
@@ -1,6 +1,17 @@
+  get "/settings", to: "types#index", as: "settings"
+
+  # items
+  resources :items, only: [:index]
```
A better pattern is:
```ruby
resources :items, only: [:index]
  collection do
    get 'search'
  end
end
```

cameronjacoby replied:
:+1: collection is not something I've used often, but I like this a lot.

cameronjacoby commented:
Changes made:
- renamed types table to style_types
- refactored clearancing_errors to be one-to-many association of clearance_batches, rather than a column on the table
- set up nested resources for clearance_batches/:clearance_batch_id/items and clearance_batches/:clearance_batch_id/clearancing_errors to handle exporting reports in more RESTful way
- refactored clearance_batch_reports_service#items_to_csv method
- removed all default_scopes
- route changes for items and style_types

Changes still to come later:
- refactoring items#search
- refactoring clearancing_service#add_items_to_clearance_batch

nickflux commented on app/decorators/item_decorator.rb:
```diff
@@ -0,0 +1,25 @@
+class ItemDecorator < Draper::Decorator
```
:+1:

nickflux commented on app/services/clearance_batch_reports_service.rb:
```diff
@@ -0,0 +1,36 @@
+      csv << headers
+      items.decorate.each do |item|
+        values = []
+        headers.each do |attr|
```
:100: :+1:
